Publish and gossip warrants

jost-s opened the pull request:
### Summary

Include warrants in publish and gossip and enable them to be fetched by the fetch module.

Add a test for each publish and gossip that confirm that warrants are published/gossiped and received by another peer. 

Queries for ops to publish have been extended to include warrants. There's a test for those.

For controlled testing of this, a conductor tuning param is added to disable warrant issuance. That prevents that invalid ops that were gossiped or published lead to a warrant on the other peer's end, who should receive the warrant from the issuer.
There's also a test that this new tuning param works.

### TODO:
- [x] CHANGELOGs updated with appropriate info
- [x] All code changes are reflected in docs, including module-level docs

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Experimental warrants are now included in publishing, gossip, and fetching when the feature is enabled, expanding ops eligible for distribution.

* **Configuration**
  * Added a tuning flag (test-only) to disable warrant issuance for testing; network-to-config translation also sets a default. Production builds unaffected.

* **Tests**
  * Feature-gated integration tests verifying warrant publishing, gossiping, fetching, and disabling issuance.

* **Documentation**
  * Changelog updated with a warrant-related entry.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jost-s commented on crates/holochain/src/core/workflow/app_validation_workflow.rs:
```diff
@@ -214,6 +214,14 @@ async fn app_validation_workflow_inner(
     #[cfg(feature = "unstable-warrants")]
     let mut warrant_op_hashes = vec![];
 
+    #[cfg(all(feature = "unstable-warrants", feature = "test_utils"))]
+    let disable_warrant_issuance = conductor
+        .config
+        .conductor_tuning_params()
+        .disable_warrant_issuance;
+    #[cfg(all(feature = "unstable-warrants", not(feature = "test_utils")))]
+    let disable_warrant_issuance = false;
```
Warrants issuance should only be disabled in tests and when the param is set to `true`.

jost-s commented on crates/holochain/tests/tests/gossip/mod.rs:
```diff
@@ -132,3 +141,156 @@ async fn new_conductor_reaches_consistency_with_existing_conductor() {
+// Publish is disabled for this test.
+#[cfg(feature = "unstable-warrants")]
+#[tokio::test(flavor = "multi_thread")]
+async fn warrant_is_gossiped() {
```
Test that warrants are published and received

jost-s commented on crates/holochain/tests/tests/publish/mod.rs:
```diff
@@ -83,3 +96,146 @@ async fn publish_terminates_after_receiving_required_validation_receipts() {
+// as he publishes it.
+#[cfg(feature = "unstable-warrants")]
+#[tokio::test(flavor = "multi_thread")]
+async fn warrant_is_published() {
```
Test that warrants are published and received

jost-s commented on crates/holochain/tests/tests/warrant_issuance.rs:
```diff
@@ -0,0 +1,73 @@
+//
+// Alice creates invalid ops, Bob receives them but should not issue any warrant.
+#[tokio::test(flavor = "multi_thread")]
+async fn warrant_issuance_can_be_disabled() {
```
Test that the tuning param to disable warrant issuance works.

ThetaSinner commented on crates/holochain/src/core/workflow/publish_dht_ops_workflow/publish_query.rs:
```diff
@@ -172,6 +203,88 @@ mod tests {
+        assert_eq!(num_to_publish, 2);
+    }
+
+    fn insert_invalid_op_warrant_op(
```
This is a little confusing to read, perhaps `insert_invalid_chain_op_warrant_op` helps a little?

ThetaSinner commented on crates/holochain/src/core/workflow/publish_dht_ops_workflow/publish_query.rs:
```diff
@@ -172,6 +203,88 @@ mod tests {
+                        - ConductorTuningParams::default().min_publish_interval(),
+                )
+                .unwrap();
+                set_receipts_complete(txn, &invalid_op_warrant.hash, false).unwrap();
```
Are validation receipts expected for warrants? I'd think not.

ThetaSinner commented on crates/holochain/src/core/workflow/publish_dht_ops_workflow/publish_query.rs:
```diff
@@ -60,6 +59,29 @@ where
+            AND
+            (DhtOp.last_publish_time IS NULL OR DhtOp.last_publish_time <= :recency_threshold)
+            AND
+            DhtOp.receipts_complete IS NULL
```
We can't rely on this to stop warrants being published, validation isn't expected to return a receipt for warrants

ThetaSinner commented on crates/holochain/src/core/workflow/publish_dht_ops_workflow/publish_query.rs:
```diff
@@ -60,6 +59,29 @@ where
+            WHERE
+            Warrant.author = :author
+            AND
+            DhtOp.withhold_publish IS NULL
```
Does not apply to warrants

ThetaSinner reviewed (commented):
Tests looking good, new config looks good. There are some inconsistencies between the data queries that is my main feedback, plus a few minor things.

I think the biggest thing that's standing out to me is that publishing of warrants won't stop. Without validation receipts, it will just keep going. I suggest we make it a one-off publish.

jost-s commented on crates/holochain/src/core/workflow/app_validation_workflow.rs:
```diff
@@ -270,9 +278,10 @@ async fn app_validation_workflow_inner(
-                if let Outcome::Rejected(_) = &outcome {
-                    let keystore = conductor.keystore();
-                    let warrant_op =
+                if !disable_warrant_issuance {
```
Alright, reduced the scope and logged the warning.

jost-s commented on crates/holochain/src/core/workflow/app_validation_workflow.rs:
```diff
@@ -354,6 +355,18 @@ async fn app_validation_workflow_inner(
 
+    // "self-publish" warrants, i.e. insert them into the DHT db as if they were published to us by another node
+    #[cfg(feature = "unstable-warrants")]
+    if warranted_ops > 0 {
```
Yep, that is redundant. Removed.

jost-s commented on crates/holochain/src/core/workflow/publish_dht_ops_workflow/publish_query.rs:
```diff
@@ -172,6 +203,88 @@ mod tests {
+
+        let agent = fixt!(AgentPubKey);
+        let db = test_authored_db();
+
+        // Insert one chain op and one warrant op into database.
+        let chain_op = create_and_insert_op(
+            &db,
+            Facts {
+                private: false,
+                within_min_period: false,
+                has_required_receipts: false,
+                is_this_agent: true,
+                store_entry: false,
+                withold_publish: false,
+            },
+            &Consistent {
+                this_agent: agent.clone(),
+            },
+        )
+        .await
+        .content;
+        let warrant_op = insert_invalid_op_warrant_op(&db, &agent).content;
+
+        let ops_to_publish = get_ops_to_publish(
+            agent.clone(),
+            &db.to_db(),
+            ConductorTuningParams::default().min_publish_interval(),
+        )
+        .await
+        .unwrap()
+        .into_iter()
+        .map(|(_, _, op)| op)
+        .collect::<Vec<_>>();
+        assert_eq!(ops_to_publish, vec![chain_op, warrant_op]);
+
+        let num_to_publish = db
+            .to_db()
+            .test_read(|txn| num_still_needing_publish(txn, agent).unwrap());
+        assert_eq!(num_to_publish, 2);
+    }
```
I've run 100 iterations of this test and took that as fixed. But it isn't, I added the ordering.

jost-s commented on crates/holochain/tests/tests/gossip/mod.rs:
```diff
@@ -132,3 +141,156 @@ async fn new_conductor_reaches_consistency_with_existing_conductor() {
+        .unwrap()
+        .into_tuple();
+
+    println!("AGENTS");
+    println!(
+        "0 alice {} url {:?}",
+        alice.agent_pubkey(),
+        conductors[0].dump_network_stats().await.unwrap().peer_urls[0]
+    );
+    println!(
+        "1 bob   {} url {:?}",
+        bob.agent_pubkey(),
+        conductors[1].dump_network_stats().await.unwrap().peer_urls[0]
+    );
+    println!(
+        "2 carol {} url {:?}",
+        carol.agent_pubkey(),
+        conductors[2].dump_network_stats().await.unwrap().peer_urls[0]
+    );
```
Yup

jost-s commented on crates/holochain/tests/tests/warrant_issuance.rs:
```diff
@@ -0,0 +1,73 @@
+        .unwrap()[0]
+        .test_read(move |txn| {
+            let store = CascadeTxnWrapper::from(txn);
+            // TODO: check_valid here should be removed once warrants are validated.
```
`check_valid` is the boolean parameter in the fn `get_warrants_for_agent`. That should be removed altogether once warrants are validated.

ThetaSinner replied:
I see! Yes okay that makes sense

jost-s commented:
Thank you for the detailed review. I've addressed your comments. Ready for review again.

I'll make another PR with extended tests for the publish query.

ThetaSinner reviewed (approved):
Super! All my feedback is addressed. I like the solution to just publish warrants once by filtering for "never published".
